src/expr/src/scalar/mod.rs

```diff
@@ -306,9 +306,15 @@ impl MirScalarExpr {
                 if expr.is_literal() {
                     *e = eval(e);
                 } else if *func == UnaryFunc::IsNull {
-                    // (<expr1> <op> <expr2>) IS NULL can often be simplified to
-                    // (<expr1> IS NULL) OR (<expr2> IS NULL).
-                    if let MirScalarExpr::CallBinary { func, expr1, expr2 } = &mut **expr {
+                    if let MirScalarExpr::Column(c) = &**expr {
+                        if !relation_type.column_types[*c].nullable {
+                            // <nonnullable column> IS NULL automatically
+                            // resolves to FALSE
+                            *e = MirScalarExpr::literal_ok(Datum::False, ScalarType::Bool);
```

Comment:
In this branch, this query has an impossible predicate in the ```Filter```

```
materialize=>  EXPLAIN  SELECT  FROM t2  AS a1  JOIN t2  AS a2  ON a1 .f1  = a2 .f2  + a2 .f1  WHERE a1 .f1  < a2 .f2  AND a1 .f1  IS  NULL  OR a2 .f2  IS  NULL   ;
                    Optimized Plan                    
------------------------------------------------------
 %0 =                                                +
 | Get materialize.public.t2 (u13)                   +
 | ArrangeBy (#0)                                    +
                                                     +
 %1 =                                                +
 | Get materialize.public.t2 (u13)                   +
 | Filter !(isnull(#0)), isnull(#0), ((#1 + #0) < #1)+
                                                     +
 %2 =                                                +
 | Join %0 %1 (= #0 (#3 + #2))                       +
 | | implementation = Differential %1 %0.(#0)        +
 | | demand = (#0)                                   +
 | Filter !(isnull(#0)), isnull(#0)                  + <- here
 | Project ()                                        +
```

The prior commit has just ```Filter !(isnull(#0))```